prober/derp.go

```diff
@@ -577,6 +602,246 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
+					return
+				}
+			}
+		}
+	}()
+
+	// Receive the packets.
+	recvC := make(chan error, 1)
+	go func() {
+		buf := make([]byte, mtu+tunStartOffset)
+		for {
+			m, err := toc.Recv()
+			if err != nil {
+				recvC <- fmt.Errorf("failed to receive: %w", err)
+				return
+			}
+			switch v := m.(type) {
+			case derp.ReceivedPacket:
+				if v.Source != fromc.SelfPublicKey() {
+					recvC <- fmt.Errorf("got data packet from unexpected source, %v", v.Source)
+					return
+				}
+				pkt := v.Data
+				copy(buf[tunStartOffset:], pkt)
+				if _, err := dev.Write([][]byte{buf[:len(pkt)+tunStartOffset]}, tunStartOffset); err != nil {
+					recvC <- fmt.Errorf("failed to write to TUN device: %w", err)
+					return
+				}
+			case derp.KeepAliveMessage:
+				// Silently ignore.
+			default:
+				log.Printf("%v: ignoring Recv frame type %T", to.Name, v)
+				// Loop.
+			}
+		}
+	}()
+
+	// Start a listener to receive the data
+	l, err := net.Listen("tcp", fmt.Sprintf("%s:0", ifAddr))
+	if err != nil {
+		return fmt.Errorf("failed to listen: %s", err)
+	}
+	defer l.Close()
+
+	randData := make([]byte, 128000)
+	_, err = crand.Read(randData)
+	if err != nil {
+		return fmt.Errorf("failed to initialize random data: %w", err)
+	}
+
+	// Read the data we've sent
+	readC := make(chan error, 1)
+	go func() {
+		readConn, err := l.Accept()
+		if err != nil {
+			readC <- err
+		}
+		defer readConn.Close()
+		_, err = io.CopyN(io.Discard, readConn, size)
+		readC <- err
+	}()
+
+	// Dial ourselves
+	_, port, err := net.SplitHostPort(l.Addr().String())
+	if err != nil {
+		return fmt.Errorf("failed to split address %q: %w", l.Addr().String(), err)
+	}
+
+	conn, err := net.Dial("tcp", fmt.Sprintf("%s:%s", destinationAddr, port))
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer conn.Close()
+
+	// Timing only includes the actual sending of data.
+	start := time.Now()
+	defer func() { transferTime.Add(time.Since(start).Seconds()) }()
+
+	// Send ourselves data
+	sendC := make(chan error, 1)
+	go func() {
+		loops := int(size) / len(randData)
+		remainderAtEnd := int(size) % len(randData)
+		totalLoops := loops
+		if remainderAtEnd > 0 {
+			totalLoops += 1
+		}
+		for i := range totalLoops {
+			b := randData
+			last := i == loops
+			if last {
+				b = randData[:remainderAtEnd]
+			}
+			if _, err := conn.Write(b); err != nil {
+				sendC <- fmt.Errorf("failed to write to conn: %w", err)
+				return
+			}
+		}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("timeout: %w", ctx.Err())
+	case err := <-tunReadC:
+		if err != nil {
+			return fmt.Errorf("error reading from TUN via %q: %w", from.Name, err)
+		}
+	case err := <-sendC:
+		if err != nil {
+			return fmt.Errorf("error sending via %q: %w", from.Name, err)
+		}
+	case err := <-recvC:
+		if err != nil {
+			return fmt.Errorf("error receiving from %q: %w", to.Name, err)
+		}
+	case err := <-readC:
+		if err != nil {
+			return fmt.Errorf("error reading from %q to TUN: %w", to.Name, err)
+		}
+	}
```

Comment:
What happens if we read from a channel other than `readC`, `err` is nil, but `readC` would return a non-nil error later?

`readC` returning nil is our signal of success: we've read all the bytes we expect from the inner TCP stream.

Reply:
we also don't wait for all the goroutines to return before returning, are we confident there are no leaking goroutines?

Reply:
My go knowledge is a bit lacking here, are you thinking we'd potentially want a [waitgroup](https://pkg.go.dev/sync#WaitGroup) (or potentially an [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup) which I think @oxtoacart mentioned in some offline discussions) to ensure we're properly waiting for the goroutines / so that we can verify that we have a non-error result for the entire group?

Reply:
I think we're okay. I tried this out by running with a large probe size (100MB) and a tiny interval (1s), and watching http://localhost:8030/debug/pprof/goroutine?debug=1.

The goroutines that read and write from the TUN device fail immediately after the timeout with either `file already closed` or `bad file descriptor` because we close the TUN device in a `defer`. The goroutine that's writing to the TCP Conn also fails because we close that connection in a `defer`. However, the goroutine that's reading from the Conn on the listening socket does not fail immediately. I'll look into it.

Reply:
Okay, we now set a read deadline on the listener's Conn to keep that goroutine from leaking past the polling interval. I've also added a WaitGroup to keep from returning until all goroutines have finished.

Reply:
> What happens if we read from a channel other than readC, err is nil, but readC would return a non-nil error later?

readC is the only channel that has non-nil errors, the other channels only report actual errors.

I've revised the naming of the channels, the comments and then handling of the results to (hopefully) make this clearer.